### ISSUE TITLE
"Risk determination not possible" - wrong info

### ISSUE THREAD

**Issue by Jo-Achim** [corona-warn-app/cwa-app-android#1081]
_I've read this topic here before - but I can't find it right now, sorry._

Android 10 with Android security update: July 1st, 2020
Nokia 7.1, Build: 00WW_4_15C_SP02
NetGuard: 2.287
Internet: WLAN (only)
CWA version: 1.2.1

Problem: "Risk determination not possible" is displayed although the risk determination is active; see screenshots.

General condition: The last CWA data exchange took place on 08/24/2020 at 7:38 am (thereafter no more internet connections until 08/26/2020).
The next time the CWA was started (with internet connection) on August 26, 2020 at around 7:08 pm / 7:09 pm, the incorrect risk determination information was displayed:
![Screenshot_20200826-190939](https://user-images.githubusercontent.com/21124877/91381912-ca625800-e828-11ea-8a1c-608b342f1c06.png)

After [->] the updated, correct data including the risk assessment were displayed:
![Screenshot_20200826-190953](https://user-images.githubusercontent.com/21124877/91381931-d77f4700-e828-11ea-8ff6-5b0f85df2306.png)

![Screenshot_20200826-191006](https://user-images.githubusercontent.com/21124877/91381940-dcdc9180-e828-11ea-94e2-24aee72d30d1.png)

Best regards.

**Comment by MikeMcC399** [comment]
Hi @Jo-Achim,

The issue is that the status "RISIKO-ERMITTLUNG AKTIV" is contradictory to the button "RISIKO-ERMITTLUNG EINSCHALTEN". If Exposure Logging is already active then there is no need to switch Exposure Logging on and the user should not be prompted to do so. That presents a confusing message to the user.

It's OK for a temporary status "Risiko-Ermittlung nicht möglich" if the app cannot pull the diagnosis keys from the server when it tries. That status was cleared and replaced by "Niedriges Risiko" after the diagnosis keys were fetched and compared to the keys on the mobile device.

**Comment by Jo-Achim** [comment]
Hi @MikeMcC399, 

yes, on the one hand I wanted to draw your attention to the fact that at least the information "switch on risk determination" when "risk determination is switched on" is wrong.

On the other hand, I believe that the message "Risk determination not possible" / "Risiko-Ermittlung nicht möglich" should not appear at all if the diagnostic keys were exchanged in the background **at the same time**. Because **in this case** the info is irrelevant. 
It would be different if no diagnostic data could be exchanged again with the current CWA call.

Best regards.

**Comment by MikeMcC399** [comment]
@Jo-Achim 

> On the other hand, I believe that the message "Risk determination not possible" / "Risiko-Ermittlung nicht möglich" should not appear at all if the diagnostic keys were exchanged in the background **at the same time**. Because **in this case** the info is irrelevant.
> It would be different if no diagnostic data could be exchanged again with the current CWA call.

I get your point. If I understood you correctly, in order to get the "Niedriges Risiko" screen to appear you had to tap on the arrow on the "Risiko-Ermittlung nicht möglich" screen. If the screen refreshed automatically (from "Risiko-Ermittlung nicht möglich" to "Niedriges Risiko") without your user action, that would be OK, but if you had to tap the arrow to get an updated status then that is not so good.

We probably have to wait for some comment from the support team / developers now for the two points raised by your issue.

**Comment by Marco2907** [comment]
Hello, @Jo-Achim thanks for your feedback. I will keep both points to our development support and will give an update as soon as possible. 

Best,
MP

Corona-Warn-App Open Source Team

**Comment by bufemc** [comment]
I have exactly the same screen, even with same date (24th August). Since days it will not work, but if I try to re-trigger it I get always this: phone is a LG G4, Android 6.0, CWA app was updated today, so version is the recent one:

![image](https://user-images.githubusercontent.com/18305092/92229468-d6de5480-eea9-11ea-81a6-6305dbcb1990.png)

**Comment by daimpi** [comment]
@bufemc
The error you're getting (39508) is a rate-limiting error, i.e. your app was trying to access a certain API functionality too frequently today. Normally what helps in such circumstances: stop the app -> wait for 24h -> open the app again.
The reason is that the limits are reset once a day, after which everything should work fine again :)

Could you try this and check whether your app can sync after this?

The tracking issue for this problem can be found here: https://github.com/corona-warn-app/cwa-app-android/issues/1021

Btw: CWA 1.3.0 has just been released in the Play-store. Maybe you want to update before you wait the 24h 🙂.

**Comment by vaubaehn** [comment]
Hi @bufemc , I'm also running an LG G4, only very rarely problems occur for me...
One possible reason in your case - are you shutting down your phone in the nights, or enable flight mode?

**Comment by vaubaehn** [comment]
@daimpi ; you can now easily find out the version of CWA by closely looking to the 39508 error message: now the 39508 is not part of the link to the FAQ anymore, so the user ends up at the top of the FAQ (introduced with #1007). But don't worry, in version 1.4.0, the 39508 will again be part of the link to the FAQ, this time in the format .../faq/#39508 (see #1074).
:laughing:

**Comment by bufemc** [comment]
@daimpi thanks, will try this to overcome "The error you're getting (39508) is a rate-limiting error".

@vaubaehn yep, this is a phone which is most times off, sometimes for days.. it's used only for particular reasons, but then this app denies to work..

for this 39508 code.. I suggest to add an explaining text then, like "limits reached" or whatever, again, for normal human beings ,)

**Comment by svengabr** [comment]
There are currently many issues opened regarding the 39508 issue.  If you want to contribute to the existing discussion, please use the main issue [here](https://github.com/corona-warn-app/cwa-app-android/issues/1021).

* [CAUSE: 3 Something went wrong. ApiException: 39508 #1021 ](https://github.com/corona-warn-app/cwa-app-android/issues/1021)
* [Resetting the CWA triggers error 39508 #934](https://github.com/corona-warn-app/cwa-app-android/issues/934)
* [Risikostatus wird nicht aktualisiert #1119](https://github.com/corona-warn-app/cwa-app-android/issues/1119)
* [Fehler bei Kommunikation mit Google API (39508) #774](https://github.com/corona-warn-app/cwa-app-android/issues/774)
* [Auf Samsung Handy seit 2 Wochen Fehlermeldung 39508 #1130](https://github.com/corona-warn-app/cwa-app-android/issues/1130)
* [Google APIException (39508) after using the SwissCovid App #816 ](https://github.com/corona-warn-app/cwa-app-android/issues/816)

Best regards,
SG

Corona-Warn-App Open Source Team

**Comment by traut21** [comment]
So when does the next update happen? Your comment is 7 days old by now.
![Screenshot_2020-09-16-06-59-11-708_de rki coronawarnapp](https://user-images.githubusercontent.com/25948981/93298481-8d6bfe80-f7f3-11ea-82ba-19e2cf4b70d0.jpg)
![Screenshot_2020-09-16-07-00-22-119_de rki coronawarnapp](https://user-images.githubusercontent.com/25948981/93298499-98bf2a00-f7f3-11ea-8c50-cf38ddc58e2c.jpg)

The error warning is annoying and does destroy trust into the proper operation of the app.

**Comment by vaubaehn** [comment]
Hi @svengabr ,

I'd like to underline, what @traut21 wrote:

> The error warning is annoying and does destroy trust into the proper operation of the app.

If you read the reviews in Google Play Store, it is quite obvious, that many people are very confused about having 'Risiko-Ermittlung aktiviert", but if risk assessment fails, a button "Risiko-Ermittlung einschalten" is presented to the user.

To sum up, what @Jo-Achim and @MikeMcC399 already pointed out perfectly: This issue here is indeed two seperate issues, deserving 2 seperate Jira tickets.

**1) The mainView/risk card is not completely refreshed, when**
- the automatic background update has not been carried out for the current day, and
- CWA is openend by the user, and
- the diagnostic keys are exchanged in the background at the same time.

In this case @Jo-Achim can see the message "Risiko-Ermittlung nicht möglich, Risiko-Ermittlung einschalten", even the update succeeded. In my case, I see a green risk card, showing last update was done yesterday.
When you then navigate to a different screen/view, no matter whether inside or outside CWA, on return/onResume, the mainView is refreshed, and the correct status is displayed in the risk card.

**Possible fix:**
- Take care, that risk card shows "updating", when ProvideDiagnosisKeys is called during a foreground process.
- Make sure, that on receiving broadcast [Action_Exposure_State_Updated](https://developers.google.com/android/reference/com/google/android/gms/nearby/exposurenotification/ExposureNotificationClient#public-static-final-string-action_exposure_state_updated) app is in foreground, mainView is active, and refresh the view.
- If the above listener is not sufficient, additionally create a listener for [Action_Exposure_Not_Found](https://developers.google.com/android/reference/com/google/android/gms/nearby/exposurenotification/ExposureNotificationClient#public-static-final-string-action_exposure_not_found) (as far as I could see from the code, no listener is set yet.) and refresh view on retrieval.

**2) The much more severe problem** imo, is, that a button "Risiko-Ermittlung einschalten" is displayed, even "Risiko-Ermittlung" is already active, up and running. This will not only happen, if there was no refresh (see 1), but also, if the automatic background update was not carried out for more than 24 hours. There are several reasons, why it was not carried out:
a) Telephone was shut down.
b) Battery optimization prevented the background task from being started.
c) Internet was closed.
d) Fetching diagnosis keys from server failed, and no re-try was set or could have been carried out.
e) The background update failed. It _failed_ means, the background task was started, the diagnosis keys are fetched from the server, and ProvideDiagnosisKeys() is called, but **ProvideDiagnosisKeys() fails**. In the past, one reason was, that the counter for resetting the counter of API call rate limit was reset at local time in CWA, but in UTC time in ENF. That problem has been fixed. The much more severe current problem is, that ProvideDiagnosisKeys() may also fail, most likely because of internal timeout for an unknown reason (slow phone, too much file operation when shifting data from app to ENF when performing 14 single calls, concurrent processes, i.e. during reboot/start-up of the phone, etc...). Most of the people, who experienced API39508 will have had a problem with timeout in any way. While the hotfix in version 1.3.1 will take account of it, in the way, that ProvideDiagnosisKeys() won't be called again by CWA in the same day after a file batch has been sent, background updates may still fail.
As the number of diagnosis keys for matching are now rising for a considerable amount due to increasing infection rates, ENF will need to perform much more work in the next weeks. Hence, the likelihood for timeout problems inside ENF is rising. As the error message API39508 will not occur anymore, in cases of ENF timeout, people will then see more often that
- automatic update has not been carried out
- the message displayed on risk card will suggest them to activate "Risiko-Ermittlung" even it is/was already activated.
To cite @traut21 again:

**Comment by vaubaehn** [comment]
Hi @GPclips , see above,
... and welcome back! :)

**Comment by traut21** [comment]
> There are several reasons, why it was not carried out:
> a) Telephone was shut down.
> b) Battery optimization prevented the background task from being started.

Concerning Battery optimization: From my radio app "VR" I know it will tell me that playback may be interrupted because of battery optimization. So there is a check for that kind of problem.

CWA never told me about potential problems because of battery optimization.

But maybe I did not get any warning because battery optimization had already been disabled.

For me it looks as if the app would not update properly if used within the background only. That's exactly what should have been fixed long ago.

**Comment by Jo-Achim** [comment]
Full agreement, especially ...
> Every installation of CWA has a small potential to save lifes by breaking infection chains.
> Every uninstalltion of CWA has the potential to contribute to the death toll.

**Comment by svengabr** [comment]
Hello everyone,

my original statement was that the 31508 errors are addressed in the upcoming hotfix release 1.3.1 however I agree with @vaubaehn that it makes sense to create two separate Jira tickets for the development team in order to counter the problems described here: https://github.com/corona-warn-app/cwa-app-android/issues/1081#issuecomment-693335476

Thank you very much for your detailed feedback!

Best regards,
SG

Corona-Warn-App Open Source Team

**Comment by svengabr** [comment]
The mainView/risk card is not completely refreshed
Internal Tracking ID: EXPOSUREAPP-2795

"Risiko-Ermittlung einschalten" is displayed, even "Risiko-Ermittlung" is already active
Internal Tracking ID: EXPOSUREAPP-2796

**Comment by bufemc** [comment]
I have to stress this again:

> The error warning is annoying and does destroy trust into the proper operation of the app.

I always have to struggle with this message, day for day, again and again, and for "normal human beings" (so, not "nerds") they will think the app stopped to work and will uninstall it. The app should be developed by putting yourself in the place of a normal person, not a nerd. It would be better to - at least - give the user "the feeling" that the app "still works". I would give the tickets a priority of urgent/high ,-)

**Comment by traut21** [comment]
It's not only the error warning.

Even if working properly, it does not update its value immediately, but does show the current update only after I take the next step to follow the (still) green "Niedriges Risiko" button area.

BTW, why doesn't the app tell me about the number of verfications? I can see that number via preferences > Google > "COVID-19 Benachrichtigungen": "Überprüfung auf mögliche Begegnungen". So it's no secret, although the App does hide that info.

**Comment by daimpi** [comment]
> BTW, why doesn't the app tell me about the number of verfications? I can see that number via preferences > Google > "COVID-19 Benachrichtigungen": "Überprüfung auf mögliche Begegnungen". So it's no secret, although the App does hide that info.

Are you referring to this number?
<img src="https://user-images.githubusercontent.com/22129968/94008971-15c64280-fda4-11ea-878a-9e721a28b9b2.png" width="400">
What do you think would be the benefits of showing this number in CWA?

Also: the number which is currently shown there is too high: https://github.com/corona-warn-app/cwa-app-android/issues/1020

**Comment by traut21** [comment]
Yes, that's the number. 

The benefit of this number would be to increase my trust in the proper operation of the CWA - that's why I asked within this topic. The current "Risk Determination not possible", the currently shown date of its last data update, all of this does increase doubt instead of trust in the proper behavior of the app.

**Comment by Jo-Achim** [comment]
I'm not sure, but there is something about the argument that cannot be dismissed out of hand. Because it shows the user that the CWA 'actually did something _and_ that there were corresponding checks'!

**Comment by daimpi** [comment]
@traut21 
> Yes, that's the number.
> 
> The benefit of this number would be to increase my trust in the proper operation of the CWA - that's why I asked within this topic. The current "Risk Determination not possible", the currently shown date of its last data update, all of this does increase doubt instead of trust in the proper behavior of the app.

hmm… maybe you could open a separate issue in the [wishlist repo](https://github.com/corona-warn-app/cwa-wishlist/issues) for this enhancement request?

**Comment by traut21** [comment]
svengabr wrote three weeks ago about a fix "the next days".

Are the 43 000 000 €, which the Telekom received for service and maintenance, used up already? 
I did not see any update or fix since.

CWA started poorly, got some fixes pretty soon, but now is in stagnation again. CWA, the next software BER?

**Comment by vaubaehn** [comment]
Hi @traut21 ,

related to https://github.com/corona-warn-app/cwa-app-android/issues/1081#issuecomment-689462301 :
> svengabr wrote three weeks ago about a fix "the next days".

unfortunately there was a misunderstanding from his side what this issue was handling about, so the announced fix didn't cover the problems discussed here. In addition, the announced fix needed to be taken back, as some unwanted side effects turned out during thorough testing.

> 
> Are the 43 000 000 €, which the Telekom received for service and maintenance, used up already?
> I did not see any update or fix since.
> CWA started poorly, got some fixes pretty soon, but now is in stagnation again. CWA, the next software BER?

I'm a user like you, and I understand your anger about unresolved issues and long delays pretty well (our poor community managers and devs can sing a song about it...), I even copied your concerns in one of the comments above => https://github.com/corona-warn-app/cwa-app-android/issues/1081#issuecomment-693335476 .
But these kind of polemic accusals towards the development team neither speed up fixing, nor motivates any of the related parties (devs, community) in a positive way, in my humble opinion.

Therefore I'd like to invite you to
- have a look behind the scenes, what other (often high priority) bugs have been fixed already and are waiting for release (after thorough testing), maybe within the next week: https://github.com/corona-warn-app/cwa-app-android/releases/tag/1.4.0-SNAPSHOT-RC5
- re-shape your anger to a more constructive way to push the concerns of the community towards the developers and project owners, to get issues with a very bad UX done first, for example:

Hi @svengabr ,
**could you do us a favor and bring up EXPOSUREAPP-2796 in your next call with the devs, please.** For the UX this issue is really bad. If this could already be fixed in 1.4.0, would be perfect!
Also EXPOSUREAPP-2795 is still an issue, I experienced it even worse when starting risk assessment in manual mode recently, but I will describe it in a seperate comment more detailed later.

Thanks to everyone!

**Comment by traut21** [comment]
Thanks for your feedback  vaubaehn. You are absolutely right, my comment was not constructive.

But I felt it was overdue. There are plenty of constructive arguments here - but I do not see any progress. 
So I felt the need to stir up some dust. Maybe this would raise the attention again for this annoying topic.

**Comment by GisoSchroederSAP** [comment]
@traut21 , guys, believe it or not, we have a full team running for the community management. So, there is no need to put an extra effort in stiring up dust, just to gain special attention.  

I'd like to refer the https://github.com/corona-warn-app/cwa-app-android/blob/dev/CODE_OF_CONDUCT.md here and kindly ask rather to stay with the discussion and with your valuable opinions regarding functions, features bugs, enhancements than with vague insinuations. 

Just to make it clear (thanks @vaubaehn for pointing to the future plans): You may not be aware of the progress on a particular issue like this one - especially, if a fix seems to be more complex than originally expected. But there are hard intense discussions about prioritization of 

- tickets and 
- bugs and 
- feature requests and 
- enhancements  and 
- text changes 

day by day. 

Not to forget: You already have the attention of the development team with this specific issue, since weeks.
So, again: We expect to get the 39508 rate limiting issue _soon_ - as it belongs to the "hot topics". Because of those discussion above, please accept, we will not offer specific dates or timelines, as those always are subject of change.

Thx.

**Comment by traut21** [comment]
> I am happy to announce that the error 39508 is well known by the development team and a fix is going to be rolled out within the next days with the hotfix release 1.3.1.

That's what was written 23 days. Yes, it's a bad idea to announce a hotfix release without knowing exactly that it will be released.

Your reply here is the first indication that "a fix seems to be more complex than originally expected" - although you do not clearly acknowledge whether this is true about the topic here or any other problem you have.

So worse than announcing a fix which does not come is not to reply and not to explain later what is going on.

If you feel it is better community management not to reply at all, I disagree.

So what's the actual problem here? Why is it that complex? Your app does stop working after some time and you have no idea why? Bummer...

**Comment by thomasaugsten** [comment]
@traut21 Fixed with v1.4 release is planned around 09.10. The issue ist we don't own the underlying API which returns the Error, we cannot control the Updates of the underlying API and we don't own the OS and OEM adjustment which calls our App in a way we cannot control.

**Comment by bufemc** [comment]
Apologies accepted ;-) However, this "bug" must at least be mitigated quickly. Or a good workaround has to be found. It confronts me every day on my LG G4.. Last time "it" worked was 23th September, that's more than a 1 week ago..

**Comment by GisoSchroederSAP** [comment]
> So worse than announcing a fix which does not come is not to reply and not to explain later what is going on.
> 
> If you feel it is better community management not to reply at all, I disagree.
> 
> So what's the actual problem here? Why is it that complex? Your app does stop working after some time and you have no idea why? Bummer...

@traut21 : Okay, to make it short:
(1) There is continous (but not immediate conversation and) reply on this topic since weeks - "not to reply" ist just wrong.
(2) The explanation for shifting the fix  was already given by @vaubaehn (thx) - the unwanted side effects during testing. At least, I hope you agree, we don't want to deliver one fix and create another new one. If you like to new more details, you may want to deep dive into the coding and into the respective API specification.
(3) "app does stop working after some time and you have no idea why? Bummer..." - this is again exactly what I don't like in the way you respond - pure assumptions, no facts, but fingerpointing. 

Again, I totally understand your frustration - you are not satisfied with the speed of fixing this particular issue, and neither am I. However, I tried to explain there are competing issues and tasks that like to allocate attention, priority and resources. 
If you cannot accept this fact, we have a community issue here.

**Comment by vaubaehn** [comment]
**TL;DR:** if you're not mentioned ;)

Hi @GisoSchroederSAP and @traut21 ,

at first, I can understand the anger of @GisoSchroederSAP . In my understanding, the developers and community mangers are in a difficult and complex "sandwich position": Sitting between stakeholders (RKI, German/European governments demanding new features like interoperability ready for deployment in the blink of an eye) and community/public (demanding new features (sometimes contrary to the stakeholder's demands) and bug fixes as soon as yesterday). If you have a look to all the finished pull requests, or you read some more deep across all issues, you may get a feeling for how much work load there must be for all involved persons, who create 'the product'. I assume, it is especially hard for the Android team, that must deal with an OS, that is often messed up by device manufacturers, or that brings in difficulties introduced by the need to rely on other open source dependencies (that bring their own bugs or unforeseen changes between (OS-)versions (this is no criticism on open source!)).
Hence, it is understandable that important fixes consume much time to deploy, or are delayed. It is also understandable that informing the community about the process is not very easy because of the work load (keeping track of many different issues) and often fast changes behind the scenes (reverting fixes because of unpredictable side effects). If, in this position, I was a community manager and being accused for slowness or whatever by the community, I would also get angry.

But on the other hand, I can clearly understand the anger and frustration expressed by @traut21 . I am sharing this anger from time to time, especially when new features are implemented while severe bugs are still persisting, leading a huge crowd of users to frustration and to uninstall CWA (according to reviews in Google Play Store). This is especially hard to realize for that sort of members of the community, who believe in the idea of digital contact tracing, and who try to support the development with passionate dedication.

What could help in this situation, imho, is more information and transparency, when community managers share more often insights out of 'the developer's premises'. As I stated above, I know it's again some hard extra work to deliver that information. There was a discussion on [Slack ](https://github.com/corona-warn-app/cwa-documentation/issues/379#issuecomment-697237851) just 2 days ago, where @cfritzsche found an issue in the backlog, that was stale for a long time, without any update on the process. In the end, @svengabr could clear up, that there is indeed a progress in that issue, but there was just no update yet. In conlusion, @svengabr is now planning to give a weekly update related to the 'hot topics'. and he started [yesterday](https://github.com/corona-warn-app/cwa-app-android/issues/933#issuecomment-702158318) (thanks a lot, @svengabr !). If this really could continue weekly, would be perfect, and I guess, it could prevent some of the frustration here.

@GisoSchroederSAP 
However, @traut21 already apologized for his wording style before:

> Thanks for your feedback [...] my comment was not constructive.

and I wonder, if it was necessary after all, to stir up the dust again by commenting in a somewhat offensive way:

> believe it or not, [...] So, there is no need to put an extra effort in stiring up dust, just to gain special attention.

I understand your anger, and given the time of your comment, it was probably coming out of an impulsive reaction. Maybe it could be better to let it go for one or two days, and then comment on these kind of issues? Also, like stated before, @traut21 already apologized. Being confronted in that harsh way after all can bring up again another harsh reaction, for sure.
And as mostly everything was already said before your comment, maybe rely a little bit more on the selfregulatory competencies of the community here :) 

@traut21 
I can also understand your anger, but your response on @GisoSchroederSAP was unfortunately not helpful at all. Again, please try to concretely ask for what you are demanding, which I understand (and share) is:
- transparency
- regular updates on processes

And to @GisoSchroederSAP and @traut21 and @svengabr and @thomasaugsten 
**This issue #1081 here is _not_ about API39508! It is about** 
1) The mainView/risk card is not completely refreshed (EXPOSUREAPP-2795)
2) "Risiko-Ermittlung einschalten" is displayed, even "Risiko-Ermittlung" is already active (EXPOSUREAPP-2796)

And (EXPOSUREAPP-2796) is indeed quite annoying, happens not only with API39508 but also in other situations (see https://github.com/corona-warn-app/cwa-app-android/issues/1081#issuecomment-693335476 ). And this flaw indeed already let many users having their CWA uninstalled, because they felt fooled (just read comments in Google Play Store in the last 3 months...)
So, @svengabr , it would really be appreciated to also get an update about this subject (sorry...)

That all being said:
@traut21 was frustrated about the delays. Also @GisoSchroederSAP said, he's frustrated about the delays. I'm also frustrated.
So, we're all in the same team. Let's don't fight each other, let's share a :beer: , and let's fight Covid togther.
Ok?

**Comment by traut21** [comment]
> @traut21 Fixed with v1.4 release is planned around 09.10. 

So then let's wait how much "around" will be stretched. I hope this release won't have to wait for the full European warn apps integration - which would be more like a software BER, I suppose.

**Comment by Carl-Orff** [comment]
> > @traut21 Fixed with v1.4 release is planned around 09.10.

Well, this issue is still not solved on 13.10. I get the error message 39508 on a regular basis since a couple of weeks. It disappears, when I choose to disable "Risikoermittlung" and wait more than 24 hours. Then the app checks out with the internet and stays green for 24 hours. 

Then, unfortunately the same error appears. My last Risiko Ermittlung is dated October 10, 04:02 hours. Since then, I get 39508 only. And the phone is connected tho my WiFi for a couple of hours each day. At least 11 hours on workdays and almost 19 hours on the weekend.

Let me put it like this: This error message ist being on the net at least since June this year. It isn't fixed. Is the app useful, when "Risiko Ermittlung nicht möglich" persists for days and can only be overcome  by applying a workaround, namely switching the "Risiko Ermittlung" off for at least 24 hours. Failure to do so will disable "Risiko Ermittlung" although it is activated. 

Being naive I think the purpose of the app is to do "Risiko Ermittlung". But it does not. Is the app useful for me? When it doesn't do its job?

And is it reasonable to think that anyone, who is not into computers, would cope with this?

**Comment by traut21** [comment]
Der Spiegel has a crushing judgment of CWA's progress: expensive, but no improvements
https://www.spiegel.de/netzwelt/web/corona-warn-app-das-teure-vergessene-mammutprojekt-a-39837ed8-adb0-4259-b98c-f5b417669efa

Maybe I should install the Irish app instead...

**Comment by vaubaehn** [comment]
> 
> Maybe I should install the Irish app instead...

Is Ireland participating in European Interoperability of tracing apps? If yes, one could consider to install Irish app and subscribe to the German diagnosis keys, if the Irish app works without errors in general.

**Comment by Carl-Orff** [comment]
> > Maybe I should install the Irish app instead...
> 
> Is Ireland participating in European Interoperability of tracing apps? If yes, one could consider to install Irish app and subscribe to the German diagnosis keys, if the Irish app works without errors in general.

If no, it is as useless as an app which works only occasionly. In this year's early July it stopped working. For more than 4 weeks. In the beginning of August, it started to work smoothly. All of a sudden.  
Now, since a couple of weeks, I get the 39508 message.

In short: In the period of 5 months the app did not work properly for at least 8 weeks. Your milage may vary.

**Comment by traut21** [comment]
It does show much more information that the CWA, but that's mainly statistics of Ireland.

It does show less information, such as when the contact tracing was updated last - which is exactly the function we do complain about here.

As far as I learned by now, the core funcion of contact tracing is not built by the app designer, but it's a core function of iOS or Android itself. The CWA is just a graphical interface on top of it.

If that's true, the app could work, regardless of an Irish or German GUI

**Comment by vaubaehn** [comment]
@Carl-Orff 
> If no, it is as useless as an app which works only occasionly. In this year's early July it stopped working. For more than 4 weeks. 

That was a Google-issue which was revealed after the first diagnosis keys in Germany were ready for downloading and matching for exposures. The java implementation of Android 7+ was not compliant to Android 6 in a special IO-function. The ENF relied on Android 7+'s implementation and was not working for all Android 6 devices. [My rough estimation](https://github.com/corona-warn-app/cwa-app-android/pull/864#issuecomment-659356206) was that around 1.000.000 devices could not make use of CWA at that time.  Luckily, public media didn't pick up that fact. Anyway, at that time, infection rates were low, and only few diagnosis keys had been shared.

> In the beginning of August, it started to work smoothly. All of a sudden.

After awesome research of @kbobrowski , Google was able to fix this. And they pushed an update of ENF in August.

> Now, since a couple of weeks, I get the 39508 message.

This is because of https://github.com/corona-warn-app/cwa-app-android/issues/1021#issuecomment-707879303 .
 
> In short: In the period of 5 months the app did not work properly for at least 8 weeks. Your milage may vary.

We can await the release of the fix soon.

**Comment by vaubaehn** [comment]
@traut21 

> As far as I learned by now, the core funcion of contact tracing is not built by the app designer, but it's a core function of iOS or Android itself. The CWA is just a graphical interface on top of it.

It's rather simplified, but not wrong. CWA needs to take of quite lots of different tasks, however.
 
> If that's true, the app could work, regardless of an Irish or German GUI

This would be correct, if the Irish app is able to subscribe to the German data (diagnosis keys). Otherwise you are stucked to the Irish diagnosis keys, which would prevent you from getting warned, unless you don't move to Ireland.

**Comment by MikeMcC399** [comment]
@traut21 
> As far as I learned by now, the core funcion of contact tracing is not built by the app designer, but it's a core function of iOS or Android itself. The CWA is just a graphical interface on top of it.

I suggest you look at the [Corona-Warn-App Solution Architecture](https://github.com/corona-warn-app/cwa-documentation/blob/master/solution_architecture.md). The CWA app is part of the German health infrastructure where the server infrastructure and links into the German virus testing labs are not provided by Google or Apple. They only provide the Exposure Notification Framework. The [Solution Architecture documentation](https://github.com/corona-warn-app/cwa-documentation/blob/master/solution_architecture.md) is heavy reading, but recommended if you want to get a better understanding of how all the parts fit together.

> If that's true, the app could work, regardless of an Irish or German GUI

If you are resident in Germany, it is not recommended to have the [Irish COVID Tracker App](https://covidtracker.gov.ie/) set as the active COVID-19 Exposure Notification app if you want to be warned about risk encounters with other people in Germany. You will not be able to warn other German users either if you are tested positive for COVID-19 by a health facility in Germany. These are functions that only work if you are using the German CWA which allows you to use German QR test codes or TANs from the German hotline.

**Comment by traut21** [comment]
> If you are resident in Germany, it is not recommended to have the [Irish COVID Tracker App](https://covidtracker.gov.ie/) set as the active COVID-19 Exposure Notification app if you want to be warned about risk encounters with other people in Germany. You will not be able to warn other German users either if you are tested positive for COVID-19 by a health facility in Germany. These are functions that only work if you are using the German CWA which allows you to use German QR test codes or TANs from the German hotline.

That's what I expected, for now. However, I do expect that a European integration is on its way.

**Comment by Carl-Orff** [comment]
That's it. Since one week  "Risiko Ermittlung" has been impossible for the App. I consider this App to be useless for me. So I uninstalled it from my phone. The App has been a story of "Etwas ist schiefgelaufen" and "Time Out" and 39508 errors.

Thanks for nothing.

**Comment by MikeMcC399** [comment]
@Carl-Orff 
>That's it. Since one week "Risiko Ermittlung" has been impossible for the App. I consider this App to be useless for me. So I uninstalled it from my phone. The App has been a story of "Etwas ist schiefgelaufen" and "Time Out" and 39508 errors.

I'm really sorry to hear about your bad experience and your frustration! 

There is a new [Version 1.5](https://github.com/corona-warn-app/cwa-app-android/releases/tag/1.5.0) on its way to the Google Play Store and we're told it should be available on Monday, Oct 19, 2020. You might want to check in here after that time and see if it has solved other peoples' problems and then consider if you want to install it.

**Comment by traut21** [comment]
So they will skip Version 1.4 -  which was due around 9.10 (of which year?).

At the moment CWA still looks like a software BER to me...

**Comment by MikeMcC399** [comment]
@traut21 
> So they will skip Version 1.4 

Yes, that is what has happened. (And before you ask, I don't know the reason for this. I'm just a user like yourself.) The app will update directly from 1.3.1 to 1.5.0 with no 1.4.x in-between.

**Comment by traut21** [comment]
Thanks, so we now got 1.5, finally. Let's see whether this problem here now is fixed, is it?

However, I now may stick to the Irish version, don't I? It now should work with German data as well.

**Comment by MikeMcC399** [comment]
@traut21 
> However, I now may stick to the Irish version, don't I? It now should work with German data as well.

We are informed that if you are tested in another country you cannot enter this test into the German app. (See https://github.com/corona-warn-app/cwa-website/issues/447#issuecomment-710079161.) So presumably the other way around, if you are using the Irish version in Germany, you cannot enter German test results into the Irish app. You would need to ask the Irish app support about this though.

**Comment by svengabr** [comment]
Could you please update to the new version 1.5 and check if the error is gone?

**Comment by Carl-Orff** [comment]
> Could you please update to the new version 1.5 and check if the error is gone?

**I** did so. Yesterday about 19:00 hours. But it is way too early to give any statements about the behavior of the error messages. I start form zero, meaning that there is still the grey box telling me "Unbekanntes Risiko".

**Comment by vaubaehn** [comment]
Hi @svengabr 

> 
> Could you please update to the new version 1.5 and check if the error is gone?

it's good to support wrt API39508 here, but just to double check and to avoid any confusion:

> **This issue #1081 here is _not_ about API39508! It is about**
> 1. The mainView/risk card is not completely refreshed (EXPOSUREAPP-2795)
> 2. "Risiko-Ermittlung einschalten" is displayed, even "Risiko-Ermittlung" is already active (EXPOSUREAPP-2796)
> 
> And (EXPOSUREAPP-2796) is indeed quite annoying, happens not only with API39508 but also in other situations (see [#1081 (comment)](https://github.com/corona-warn-app/cwa-app-android/issues/1081#issuecomment-693335476) ). And this flaw indeed already let many users having their CWA uninstalled, because they felt fooled (just read comments in Google Play Store in the last 3 months...)
> So, @svengabr , it would really be appreciated to also get an update about this subject (sorry...)

And EXPOSUREAPP-2795 as well as EXPOSUREAPP-2796 have not yet been solved with 1.5.0 - so this issue should be kept open :angel:

**Comment by traut21** [comment]
> > Could you please update to the new version 1.5 and check if the error is gone?
> 
> **I** did so. Yesterday about 19:00 hours. But it is way too early to give any statements about the behavior of the error messages. I start form zero, meaning that there is still the grey box telling me "Unbekanntes Risiko".

Yes, up to now it just shows e.g. the green box "Aktualisiert: Gestern, 6:31". Only when I click the green area, it does provide the update "Aktualisiert: Heute, 13:41"

So this does not give me new trust yet.
1) it does not update within the background more frequently than every 24 hours
2) it does not show the updated and current value yet, but does show me the outdated value only.

I think 2) UI is a poor choice, but even worse it's only that way that I do notice 1)

But that's a very different issue - is this reported within another issue yet?

**Comment by ndegendogo** [comment]
> it does not update within the background more frequently than every 24 hours

@traut21 you mean [this](https://github.com/corona-warn-app/cwa-backlog/issues/2)?

Feel free to upvote.

**Comment by daimpi** [comment]
@traut21 
>  Only when I click the green area, it does provide the update "Aktualisiert: Heute, 13:41"

When you go back to the main screen in CWA after that, does it then also show "Aktualisiert: Heute, 13:41" there or still "Aktualisiert: Gestern, 6:31"?

**Comment by traut21** [comment]
> When you go back to the main screen in CWA after that, does it then also show "Aktualisiert: Heute, 13:41" there or still "Aktualisiert: Gestern, 6:31"?

It then does show Aktualisiert: Heute, 13:41, as expected

**Comment by daimpi** [comment]
@traut21 
Ok, thanks. Iiuc you're currently not sure whether the update is only working after you're tapping on the green area or whether this is "just" a display error, right?

Could you tomorrow try the following: 
1. Open CWA and see whether it shows an outdated check timestamp
2. If this is the case: **don't** tap on the green area, instead go into the Google settings on your phone, go to Covid-19 Notifications and look when the last check has been performed. If the check has been performed on the current day, then it's "just" a displaying error but the app is working fine. If not: there is a more serious underlying error.

I would guess that it's the former (i.e. "just" a displaying error) but it would be interesting to know for sure 🙂.

**Comment by Jo-Achim** [comment]
@daimpi  
I think I can answer your question as a hint (because I cannot judge any overlaps in time); see screenshots.

![Screenshot_20201021-081940](https://user-images.githubusercontent.com/21124877/96681680-f78a3d00-1377-11eb-828f-c9a1aff38433.png)

![Screenshot_20201021-082047](https://user-images.githubusercontent.com/21124877/96681723-0670ef80-1378-11eb-9f61-0dd1aab2f193.png)


After taking the screenshot above and switching back to the CWA ...

![Screenshot_20201021-082600](https://user-images.githubusercontent.com/21124877/96681747-11c41b00-1378-11eb-9039-b31db174738b.png)

Best regards.

**Comment by daimpi** [comment]
@Jo-Achim thanks 🙂 

This would confirm this being a displaying error. Is it correct, that the first screen is staying for a significant amount of time/forever if you don't switch contexts?

This means 
> The mainView/risk card is not completely refreshed (EXPOSUREAPP-2795)

is still present on CWA 1.5 as [expected](https://github.com/corona-warn-app/cwa-app-android/issues/1081#issuecomment-712790536).

**Comment by Jo-Achim** [comment]
@daimpi 
> 
> This would confirm this being a displaying error. Is it correct, that the first screen is staying for a significant amount of time/forever if you don't switch contexts?
> 
> This means
> 
> > The mainView/risk card is not completely refreshed (EXPOSUREAPP-2795)
> 
> is still present on CWA 1.5 as [expected](https://github.com/corona-warn-app/cwa-app-android/issues/1081#issuecomment-712790536).
>
Yes, I think it's a display error - but it doesn't always occur. So there are likely to be other conditions that lead to "Updated" ("Aktualisiert: ...") being displayed correctly or incorrectly immediately after opening the CWA.
And from what I have observed, a false display will remain false as long as no other action is taken. (Tap to "Your Risk Status" or, like today, looking at "Covid-19 Notifications" and then back to CWA; see screenshots from today.)

As far as I can remember, it happened in _earlier_ CWA versions (<1.5) that the wrong "Updated" ("Aktualisiert") was also displayed there after tapping to "Your risk status"; only the return to the start screen corrected both displays.

PS: The screenshots above are from today, for example. And yes, I have CWA 1.5 installed.

**Comment by dsarkar** [comment]
Hi @Jo-Achim,

thanks for your contribution and detailed explanation. We will contact the developers and new updates on EXPOSUREAPP-2795 will be posted here hopefully shortly. 

Best wishes
DS

Corona-Warn-App Open Source Team

**Comment by Jo-Achim** [comment]
@daimpi, referring to the above post https://github.com/corona-warn-app/cwa-app-android/issues/1081#issuecomment-713344103 maybe another hint.
Today after the first call of the CWA (v 1.5.0) "Updated: Yesterday ..." was displayed again. When I wanted to take a screenshot, I pressed the on / off / lock button too briefly ...
After switching on again, "Updated: Today 08:2**3**" was correctly displayed in the CWA (start screen). Interestingly, under "Covid-19 notification", the time of today's review is given as 08:2**4**.

![Screenshot_20201024-082446](https://user-images.githubusercontent.com/21124877/97070316-aaa4a180-15d7-11eb-9333-ff67198485eb.png)

![Screenshot_20201024-082502](https://user-images.githubusercontent.com/21124877/97070317-ab3d3800-15d7-11eb-9412-2a1caa6c6f9d.png)

Best regards.

**Comment by daimpi** [comment]
@Jo-Achim thanks for the update 🙂 

I think what you were observing is: https://github.com/corona-warn-app/cwa-app-android/issues/948

I don't necessarily think there is a connection with this issue here… Imho it's more likely that it's just a rounding error of the timestamp.

**Comment by traut21** [comment]
I'm very sorry to report that the error has not been fixed
![Screenshot_2020-10-25-10-54-09-917_de rki coronawarnapp](https://user-images.githubusercontent.com/25948981/97106831-195f2900-16c4-11eb-9da4-bf2a93de39b4.jpg)
![Screenshot_2020-10-25-10-54-31-672_de rki coronawarnapp](https://user-images.githubusercontent.com/25948981/97106833-1c5a1980-16c4-11eb-8d51-62d98f6f1f92.jpg)
I had WLAN most of the time every day, but the CWA failed to look for updates for several days.

**Comment by daimpi** [comment]
@traut21 are we "just" dealing with the displaying problem in your case?
> The mainView/risk card is not completely refreshed (EXPOSUREAPP-2795)

Or did CWA actually not perform any updates even when you opened it manually?

**Comment by Jo-Achim** [comment]
@traut21, allow me a little comment.
I usually only use WiFi for data exchange. If the cost aspect should play a role for you in this regard, see CWA FAQs: "If I use the app outside of a WLAN, will I have to pay for the transfer of mobile data?" / "Wenn ich die App außerhalb eines WLANs nutze, kommen dann Kosten für die Übertragung mobiler Daten auf mich zu?".
Maybe it's a workaround for you !?

**Comment by traut21** [comment]
> @traut21 are we "just" dealing with the displaying problem in your case?
> 
> > The mainView/risk card is not completely refreshed (EXPOSUREAPP-2795)
> 
> Or did CWA actually not perform any updates even when you opened it manually?

when I open it manually, it does show that it updated. But the message tells me that it did never do the update in the background since the last time I had opened it.

So the risk determination still does not work as needed.

**Comment by traut21** [comment]
> @traut21, allow me a little comment.
> I usually only use WiFi for data exchange. If the cost aspect should play a role for you in this regard, see CWA FAQs: "When I'm not connected to a Wi-Fi network, will I be charged for any mobile data transmitted via the app?" / "Wenn ich die App außerhalb eines WLANs nutze, kommen dann Kosten für die Übertragung mobiler Daten auf mich zu?".
> Maybe it's a workaround for you !?

I don't think so. I do have both WLAN and mobile data turned on. But the CWA failed to obtain any updates.

**Comment by daimpi** [comment]
@traut21 
> I don't think so. I do have both WLAN and mobile data turned on. But the CWA failed to obtain any updates.

Do I understand you correctly that this is not just a displaying error? It persists even when you close and re-open CWA?

Some more questions:
- do you observe any other Error messages
- could you share your EN log
- which device do you use with which Android version?
- which CWA version are you running?
- which [ENF version](https://www.coronawarn.app/en/faq/#ENF_version) are you running?

**Comment by traut21** [comment]
If it tells me that the last update was on 2020-10-21, I have no doubt about that. But this means that it did not do any update since I did it last time, manually.
The status Risk determination not possible then is shown automatically, after about more than 48 hours since the last update?

I don't know about the close and re-open because I now had performed a manual update.

If the CWA does work with manual updates only, that's not what I need.

- no error messages
- where do I find the EN log?
- Android 10 on a Redmi 9
- CWA 1.5.0
- ENF 17203915000

**Comment by daimpi** [comment]
@traut21 thanks for the info 🙂.
> I don't know about the close and re-open because I now had performed a manual update.

How do you perform a manual update? What steps do you take besides opening CWA if any?

> where do I find the EN log?

On your phone's main screen: settings -> Google -> COVID-19 exposure notifications
click on the field where it says "checks performed in the last 14 days", this will bring you to a long list. You can export this list by clicking on the burger menu in the top right corner of the screen and select export.
To share the file here on GH just rename the file-ending from `.json` to `.txt`.

**Comment by traut21** [comment]
> @traut21 thanks for the info 🙂.
> 
> > I don't know about the close and re-open because I now had performed a manual update.
> 
> How do you perform a manual update? What steps do you take besides opening CWA if any?

I do click on the usally green, this time white area.

> > where do I find the EN log?
> 
> On your phone's main screen: settings -> Google -> COVID-19 exposure notifications
> click on the field where it says "checks performed in the last 14 days", this will bring you to a long list. You can export this list by clicking on the burger menu in the top right corner of the screen and select export.

ok, It does confirm no update between 25. and 21.10.
> To share the file here on GH just rename the file-ending from `.json` to `.txt`.

**Comment by daimpi** [comment]
@traut21 thanks 🙂.
> I do click on the usally green, this time white area.

I do think I know now what is happening: this seems to be a combination of issue https://github.com/corona-warn-app/cwa-app-android/issues/933 and 
> "Risiko-Ermittlung einschalten" is displayed, even "Risiko-Ermittlung" is already active
Internal Tracking ID: EXPOSUREAPP-2796

Do you have Prioritized Background Activity (PBA) enabled in the CWA settings? If so: could you leave a short comment over at #933 describing your situation?

Additionally you might find the following instructions helpful: https://dontkillmyapp.com/xiaomi

**Comment by traut21** [comment]
> Do you have Prioritized Background Activity (PBA) enabled in the CWA settings? If so: could you leave a short comment over at #933 describing your situation?

PBA is enabled

> Additionally you might find the following instructions helpful:
> 
>     * https://www.coronawarn.app/en/faq/#no_risk_update
> 
>     * https://dontkillmyapp.com/xiaomi

I expected hat 1.5 should fix what has been reported in August here as a bug. Bummer, still not working...

**Comment by traut21** [comment]
I'll need to watch it for some more time. But at the moment CWA 1.5.1 seems to update well within the background on my Redmi 9.

**Comment by traut21** [comment]
I'm sorry to report that the bug has not yet been fixed.

CWA today showed to me the common error message
![Screenshot_2020-11-13-08-45-45-303_de rki coronawarnapp](https://user-images.githubusercontent.com/25948981/99042440-0508a080-258d-11eb-860c-0e062e3fff5b.jpg)
which changed only after I went for a refresh
![Screenshot_2020-11-13-08-46-32-726_de rki coronawarnapp](https://user-images.githubusercontent.com/25948981/99042448-08039100-258d-11eb-89b1-8fc952285647.jpg)

**Comment by d4rken** [comment]
Is this still an issue in 1.9.1+?

> "Risiko Ermittlung einschalten" 

There should be only 4 cards now.
* Increased risk (RED) <- Reflects last task result
* Low risk (GREEN) <- Reflects last task result
* Risk failed (WHITE) <- Reflects last task result
* Tracing off (WHITE) <- Reflects current device status

So if tracing is disabled, the "Tracing disabled" card with an "Enable tracing" button is shown, and in all other cases the card itself reflects the last result of the [RiskLevelTask](https://github.com/corona-warn-app/cwa-app-android/blob/715c30ad482301594b77a003ce870f5997302541/Corona-Warn-App/src/main/java/de/rki/coronawarnapp/risk/RiskLevelTask.kt#L34).
Which will cause it to either be red/green, or white [if the task failed](https://github.com/corona-warn-app/cwa-app-android/blob/5bfc89c1d91bf8afd6aa610dbd611b4515d30e46/Corona-Warn-App/src/main/java/de/rki/coronawarnapp/risk/RiskLevelResult.kt#L52-L58).

> Timestamps

The "last updated" timestamp as of 1.9.1 is "the last time we submitted keys to the ENF":
https://github.com/corona-warn-app/cwa-app-android/blob/5bfc89c1d91bf8afd6aa610dbd611b4515d30e46/Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/tracing/card/TracingCardStateProvider.kt#L67

while this may be off by a few minutes from what Google shows in the system's ENF, I don't have a better solution at the moment. I've tried "time when we submit to the ENF" and "time when the ENF tells us it's done calculating" and neither consistently matched the system's ENF timestamps.

**Comment by dsarkar** [comment]
Hi @traut21, @Jo-Achim, please see https://github.com/corona-warn-app/cwa-app-android/issues/1081#issuecomment-750344065.
 
Best wishes, DS

---
Corona-Warn-App Open Source Team
